pallets/foreign-investments/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+	/// tranche tokens burned for the conversion into pool currency based on the
+	/// fulfillment price(s) during collection.
+	///
+	/// NOTE: The lifetime of this storage starts with collecting a redemption
+	/// in pool currency and ends with having swapped the entire amount to
+	/// return currency which is assumed to be asynchronous.
+	#[pallet::storage]
+	pub(super) type CollectedRedemptionTrancheTokens<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Blake2_128Concat,
+		T::InvestmentId,
+		T::Balance,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		ForeignInvestmentUpdated {
+			investor: T::AccountId,
+			investment_id: T::InvestmentId,
+			state: InvestState<T::Balance, T::CurrencyId>,
+		},
+		ForeignInvestmentCleared {
+			investor: T::AccountId,
+			investment_id: T::InvestmentId,
+		},
+		ForeignRedemptionUpdated {
+			investor: T::AccountId,
+			investment_id: T::InvestmentId,
+			state: RedeemState<T::Balance, T::CurrencyId>,
+		},
+		ForeignRedemptionCleared {
+			investor: T::AccountId,
+			investment_id: T::InvestmentId,
+		},
+	}
+
+	#[pallet::error]
+	// TODO: Add more errors
+	pub enum Error<T> {
+		InvalidInvestmentCurrency,
+		/// Failed to retrieve the `TokenSwapReason` from the given
+		/// `TokenSwapOrderId`.
+		InvestmentInfoNotFound,
+		/// Failed to retrieve the `TokenSwapReason` from the given
+		/// `TokenSwapOrderId`.
+		TokenSwapReasonNotFound,
+		/// Failed to transition the `InvestState`.
+		InvestError(InvestError),
+		/// Failed to transition the `RedeemState.`
+		RedeemError(RedeemError),
+	}
+
+	impl<T> From<InvestError> for Error<T> {
+		fn from(error: InvestError) -> Self {
+			Error::<T>::InvestError(error)
+		}
+	}
+
+	impl<T> From<RedeemError> for Error<T> {
+		fn from(error: RedeemError) -> Self {
+			Error::<T>::RedeemError(error)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Attempts to transition an `InvestState` after an epoch execution:
+		/// * If the state includes `InvestmentOngoing` and the unprocessed
+		///   investment amount is zero, removes `InvestmentOngoing`.
+		/// * If the state includes `InvestmentOngoing` and the unprocessed
+		///   investment amount is positive, updates the `invest_amount` to the
+		///   unprocessed one.
+		///
+		/// NOOP: If the unprocessed investment amount is zero or the state does
+		/// not include `InvestmentOngoing`
+		#[pallet::weight(100_000_000 + T::DbWeight::get().reads_writes(5, 5).ref_time())]
+		#[pallet::call_index(1)]
+		pub fn nudge_invest_state(
+			origin: OriginFor<T>,
+			investor: T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			if let Some(invest_state) = InvestmentState::<T>::get(&investor, investment_id) {
+				let amount_unprocessed_investment =
+					T::Investment::investment(&investor, investment_id)?;
```

Comment:
This amount might contain be below the actual `amount_unprocessed_investment ` amount. The reason being is, that a user needs to collect before this amount reflects the actual `amount_unprocessed_investment`. 

Is this okay?

Otherwise we need to
* execute `collect` in `rame_support::storage::with_storage_layer` if we want the state change to be reverted that collect is introducing
* execute `collect` 

Note: In both cases we must check that we are `FullyCollected`.

Reply:
Since the file is outdated, I cannot see to which function this snippet belongs. I assume it did to `collect_foreign_investment`. Now we are calling `T::Investment::investment(who, investment_id)` after collecting the investment. So we should be fine?

https://github.com/centrifuge/centrifuge-chain/blob/3606df98063e872f277d2dfd176153b5cf9fffd1/pallets/foreign-investments/src/impls/mod.rs#L233